### PR TITLE
Refactor name setting for complex cases

### PR DESCRIPTION
**Pull request by yiraeChristineKim**
This PR supports the more complex cases of setting names. Especially, `consolidateManifests` is true and the `manifest-path name` is present.
https://issues.redhat.com/browse/ACM-1256

### CODE REVIEW

File: internal/ordering_test.go [hunk 1]
```diff
@@ -293,10 +293,16 @@ metadata:
...
   orderPolicies: true
   namespace: my-policies
-  consolidateManifests: false
+  consolidateManifests: true
 policies:
 - name: one
   manifests:
+  - path: {{printf "%v/%v" .Dir "configmap.yaml"}}
+    name: tiger
+  - path: {{printf "%v/%v" .Dir "configmap.yaml"}}
```
Comment (yiraeChristineKim) [inline internal/ordering_test.go:302]: Although this `manifest path name` has "rabbit", this name is ignored because it is consolidateManifests. The configuration name will be "tiger"

File: internal/utils.go [hunk 2]
```diff
@@ -326,23 +329,31 @@ func getPolicyTemplates(policyConf *types.PolicyConfig) ([]map[string]interface{
 // When index is 0, it won't attach any number. When useManifestName is true,
 // it returns the policyConf.Manifests[manifestGroupIndex].Name otherwise,
 // It returns policyConf.Name + index
-func getConfigurationPolicyName(policyConf *types.PolicyConfig, manifestGroupIndex, policyNum,
-	manifestIndex int, useManifestName bool,
+func getConfigurationPolicyName(policyConf *types.PolicyConfig, manifestGroupIndex int,
+	usedPolicyNameNum *int, usedManifestNameNum *int,
```
[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply internal/utils.go:333]: It is related to previous PR sorry about confusing

[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply internal/utils.go:333]: Inside the function, we don't know which name is used so

File: internal/ordering_test.go (hunk 1 above)
Comment (yiraeChristineKim) [reply internal/ordering_test.go:302]: I understand I wanted to cover most of the edge cases. because base cases are covered in other unit tests

Comment (yiraeChristineKim) [reply internal/ordering_test.go:302]: I can say this is the **ULTIMATE** test case.

Comment (yiraeChristineKim) [reply internal/ordering_test.go:302]: Maybe it is good to split

File: internal/utils.go (hunk 2 above)
Comment (dhaiducek) [inline internal/utils.go:333]: (optional) Combine the arg types; Use proposed name
```suggestion
	policyNameCounter, manifestNameCounter *int,
```

File: internal/utils.go [hunk 3]
```diff
@@ -326,23 +329,31 @@ func getPolicyTemplates(policyConf *types.PolicyConfig) ([]map[string]interface{
 // When index is 0, it won't attach any number. When useManifestName is true,
 // it returns the policyConf.Manifests[manifestGroupIndex].Name otherwise,
 // It returns policyConf.Name + index
```
Comment (dhaiducek) [inline internal/utils.go:331]: The description for the function needs an update. I was trying to think about how it works, so I drafted one as a starter:
```golang
// getConfigurationPolicyName returns the `ConfigurationPolicy` name based on the provided PolicyConfig.
// If `consoldationManifests` is true or `manifests[].name` is not provided, it falls back to `policies[].name`.
// When a name is assigned, it increments the associated counter, using it as a suffix if it's greater
// than one to create unique names. The `manifestNameCounter` can be set to nil to force the function to
// return `policies[].name` regardless of whether `manifests[].name` is set.
```

**Review by dhaiducek — COMMENTED** [review]
Looks good! The counter is definitely tricky. I've left some comments to consider.

I also keep going back and forth in my mind on whether we should adjust the current numbering system. This is how it should have been implemented originally, but it is a change in behavior and, if we're going to implement this, it shouldn't be backported.

File: internal/utils.go [hunk 4]
```diff
@@ -326,23 +329,31 @@ func getPolicyTemplates(policyConf *types.PolicyConfig) ([]map[string]interface{
...
+	}
+
+	if usedManifestNameNum != nil && configName != "" {
+		*usedManifestNameNum++
+		if *usedManifestNameNum > 1 {
+			return fmt.Sprintf("%s%d", configName, *usedManifestNameNum)
 		}
 
 		return configName
 	}
```
[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply internal/utils.go:350]: Should we add error message or warning message for this, otherwise this is really confusing usesr.

[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply internal/utils.go:350]: Brian commented https://issues.redhat.com/browse/ACM-12563

File: internal/utils.go [hunk 5]
```diff
@@ -177,7 +179,9 @@ func getPolicyTemplates(policyConf *types.PolicyConfig) ([]map[string]interface{
 		ignorePending := policyConf.Manifests[i].IgnorePending
 		extraDeps := policyConf.Manifests[i].ExtraDependencies
 
-		for j, manifest := range manifestGroup {
+		usedManifestNameNum := 0
```
[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply internal/utils.go:182]: I love love love you suggestion

File: internal/utils.go [hunk 6]
```diff
@@ -169,6 +169,8 @@ func getPolicyTemplates(policyConf *types.PolicyConfig) ([]map[string]interface{
 	objectTemplates := make([]map[string]interface{}, 0, objectTemplatesLength)
 	policyTemplates := make([]map[string]interface{}, 0, policyTemplatesLength)
 
+	usedPolicyNameNum := 0
```
[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply internal/utils.go:172]: love love suggestion

File: internal/utils.go (hunk 3 above)
Comment (yiraeChristineKim) [reply internal/utils.go:331]: love it

File: docs/policygenerator-reference.yaml [hunk 7]
```diff
@@ -236,6 +236,8 @@ policies:
       - path: ""
         # Optional. This name is used when ConsolidateManifests is set to false and will serve as the ConfigurationPolicy name.
         # If multiple manifests are present in the path, an index number will be appended.
+        # If multiple manifests are present and their names are provided, with `consolidated` set to true,
+        # the name of the first manifest will be used for all manifest paths.
```
[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply docs/policygenerator-reference.yaml:240]: I agree with Brian. If the user prefers to use `policies.name`, they can leave `manifests[num].name` empty. However, `manifests[num].name` provides more flexibility and options for the user, which is why I prefer it, even when consolidate=true. Could you elaborate on your thoughts or raise any concerns you might have?

File: internal/utils.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dhaiducek) [reply internal/utils.go:350]: I don't think it's confusing for the user since the manifests are being combined so it's using the parent configuration, and I _think_ (?) this PR resolves Brian's code bug since the `manifest[].name` is handled separately for `object-templates-raw`.

[the remark this answers is not in the document]
Comment (dhaiducek) [reply internal/utils.go:350]: Oh, I see. I wasn't thinking about this allowing configuration of the ConfigurationPolicy to have a different name than the Policy. Yes, this makes sense then to just use the first name it finds.

[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply internal/utils.go:350]: So leave it as t is?

[the remark this answers is not in the document]
Comment (dhaiducek) [reply internal/utils.go:350]: Leave the logic, yes. I still think the conditionals could optionally be combined:
```golang
// The manifest name is provided
if configName != "" {
	// The manifest name counter is not nil so must be incremented and used if it's greater than one
	if usedManifestNameNum != nil {
		*usedManifestNameNum++
		if *usedManifestNameNum > 1 {
			return fmt.Sprintf("%s%d", configName, *usedManifestNameNum)
		}
	}

	// Return the manifest name without a suffix
	return configName
}
```
I might have to think more about whether we should be doing duplicate name validation, but right now I'm thinking that can be chalked up to user error since the framework would complain once it's deployed.

[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply internal/utils.go]: I am not sure this repos has sonarcloud. Sonar doesn't like too many if-if child.

File: docs/policygenerator-reference.yaml [hunk 8]
```diff
@@ -236,6 +236,8 @@ policies:
       - path: ""
         # Optional. This name is used when ConsolidateManifests is set to false and will serve as the ConfigurationPolicy name.
         # If multiple manifests are present in the path, an index number will be appended.
+        # If multiple manifests are present and their names are provided, with `consolidated` set to true,
```
Comment (mprahl) [inline docs/policygenerator-reference.yaml:239]: Did you mean this?
```suggestion
        # If multiple manifests are present and their names are provided, with `consolidateManifests` set to true,
```

File: internal/utils.go [hunk 9]
```diff
@@ -282,11 +285,13 @@ func getPolicyTemplates(policyConf *types.PolicyConfig) ([]map[string]interface{
 	// just build one policyTemplate by using the above non-empty consolidated objectTemplates
 	// ConsolidateManifests = true or there is non-policy-type manifest
 	if policyConf.ConsolidateManifests && len(objectTemplates) > 0 {
+		// ConsolidateManifests is true and multiple manifest[].names are provided, the configuration policy
```
Comment (mprahl) [inline internal/utils.go:288]:
```suggestion
		// If ConsolidateManifests is true and multiple manifest[].names are provided, the configuration policy
```

**Review by mprahl — APPROVED** [review]

**Review by dhaiducek — APPROVED** [review]
There are some edge cases here where the same name could be specified throughout for `policies[].name` and `manifests[].name`.

Though for `policies[].name`, it was a blocking thing since the object couldn't be deployed at all so it really needed to be an error case. For `manifests[].name`, the framework would let the user know that the same name is in use so we could chalk it up to user error and not have it be blocking, which I think I'd be fine with.

I still maintain that the numbering changes should not be backported. If there is a backport, it should only include the portion of this PR dealing with `object-templates-raw`.

**Comment by yiraeChristineKim** [comment]
> There are some edge cases here where the same name could be specified throughout for `policies[].name` and `manifests[].name`.
> 
> Though for `policies[].name`, it was a blocking thing since the object couldn't be deployed at all so it really needed to be an error case. For `manifests[].name`, the framework would let the user know that the same name is in use so we could chalk it up to user error and not have it be blocking, which I think I'd be fine with.
> 
> I still maintain that the numbering changes should not be backported. If there is a backport, it should only include the portion of this PR dealing with `object-templates-raw`.
The Fix version is 2.12 so we don't need to backport Thanks for your comment.  I haven't thought about a "same name" case. Good Catch!! 
Should we give the use a warning message when manifests[].name and policies[].name are same?
